### PR TITLE
Prometheus exporter fix for v2.x

### PR DESCRIPTION
**Pull request by lukas-vlcek**
# Prometheus exporter fix for v2.x

This PR brings fixes for several issues in existing Prometheus exporter plugin implementation. **It brings fixes into `2.x` branch** and all the following text is relevant mostly to this branch although I think similar series of fixes is needed for `master` branch as well (this can be addressed by followup PR depending on feedback from plugin maintainers).

## Why I did this?

My main goal was to make sure Prometheus exporter plugin can work well with Elasticsearch cluster that is using [SearchGuard plugin](http://floragunn.com/searchguard) for security. As of writing I was focusing on Elasticsearch 2.x but my interest going forward is to make sure Prometheus exporter plugin will work well with later ES versions (5.x, ...) too.

While I was doing my evaluation with Elasticsearch 2.x and SearchGuard I hit couple of issues (see below for details). I realized that fixing all those issues is easier with one monolithic PR rather than several independent PRs. Part of the story was that changes required significant architecture reimplementation.

### [#57] thread context handling is buggy

[This issue](https://github.com/vvanholl/elasticsearch-prometheus-exporter/issues/57) is caused by the fact that `Client` instance is shared and reused for other REST calls. In connection with SearchGuard plugin this represents major security issue (and I think the same can apply to other security solutions, this may not be SG exclusive). Please follow up the [#57](https://github.com/vvanholl/elasticsearch-prometheus-exporter/issues/57) for further details.

The solution is to not keep and reuse `client` instance.

### [#58] Do not execute a blocking operation on a networking thread

[This issue](https://github.com/vvanholl/elasticsearch-prometheus-exporter/issues/58) actually stems from the fact that plugin is using blocking client calls in the REST layer. In normal situation users are not hit by this issue because `Asserts` are not enabled. However, when running ES with `Assets` enabled (like when implementing integration tests using [Elasticsarch test framewrok](https://www.elastic.co/guide/en/elasticsearch/reference/2.4/testing-framework.html)) the execution leads to exception.

The correct solution is to move the plugin logic to lower layers in Elasticsearch. From REST layer to Transport layer and use strictly async calls. At the same time it is very important to make sure that specific actions are executed locally on the node that receives the initial REST call. This PR brings all of this.

One of the tradeoffs here is that for each REST call a new instance of `PrometheusMetricsCollector` is created. But I do not think this is really expensive especially given Prometheus server is typically setup to scrape metrics once per 10 seconds or so. May be even there is a way how we can save some cycles and refactor & reuse objects that handle individual metrics registeration. 

### [new feature] Possibility to control access to node REST API with SearchGuard

One of the benefits of moving the logic from REST to Transport layer is the possibility to implement Elasticsearch `Action` which enables much better **access control to plugin REST API** (`/_prometheus/metrics`) when using SearchGuard.

Right now users accessing REST API have to have permissions for all behind-the-scene calls. Specifically, they have to be granted permissions for **cluster health** and **node stats** (and possibly others for `5.x` branch). In SearchGuard config it looks like this:

````yml
sg_role_both_perms:
  cluster:
    - cluster:monitor/health
    - cluster:monitor/nodes/stats
````
If user is missing any of these then the REST call fails. On the other hand, any user that has these perms is allowed.

This PR intorduces new permission for **prometheus metrics** REST API only. In SearchGuard it is now possible to do this:

````yml
sg_role_prometheus_perm:
  cluster:
    - cluster:monitor/prometheus/metrics
````
All users having this single permission are allowed in and no other permission is required. This also means that if in the future the prometheus plugin will start accessing data from other resources in Elasticsearch the configuration for SearchGuard does not need to change.

I did test this feature using the same code that I used to demonstrate [#57]. That repo now contains two new branches [`case_1_fixed`](https://github.com/lukas-vlcek/es-sg-prometheus-integration/tree/case_1_fixed) and [`case_2_fixed`](https://github.com/lukas-vlcek/es-sg-prometheus-integration/tree/case_2_fixed) and **both are using custom build of fixed Prometheus exporter plugin**. It can bee seen in associated Travis logs ([here for case\_1\_fixed](https://travis-ci.org/lukas-vlcek/es-sg-prometheus-integration/builds/287850949#L2223) and [here for case\_2\_fixed](https://travis-ci.org/lukas-vlcek/es-sg-prometheus-integration/builds/287852392#L2223)) that both cases now work correctly.

### [#63] Why using node name instead of node ID for label?

Strictly speaking, this is a **new feature** rather than an issue. See [#63](https://github.com/vvanholl/elasticsearch-prometheus-exporter/issues/63) for details.

### [new feature] Introduction of integration tests

I added integration tests. I think they are self explanatory. One of the key aspectes they test is the fact that Prometheus exporter returns data only for the node that received the HTTP call (the nodes stats).

Use `mvn clean verify` to run both **unit tests** and **integration tests**.

### Anything else...?

While I was on it I did some other minor updates here and there. For example I updated the `java_client` version and did some improvements in logging.

## Where to go from here?

- I would love to hear back
- I will be happy to answer any questions
- I want to help get the same fixes to other branches
- I would like to suggest adding integration with Travis (now that integration tests are in place it makes a lot of sense)

### CODE REVIEW

**Comment by lukas-vlcek** [comment]
@vvanholl @davidkarlsen any comments or concerns on this please?

**Comment by vvanholl** [comment]
Hi @lukas-vlcek, I just want to notify you I am currently reading your PR

File: src/main/java/org/elasticsearch/rest/action/prometheus/RestPrometheusMetricsAction.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+            public RestResponse buildResponse(NodePrometheusMetricsResponse response) throws Exception {
+
+                String clusterName = response.getClusterHealth().getClusterName();
+                String nodeName = response.getNodeStats().getNode().getName();
+                String nodeId = response.getNodeStats().getNode().getId();
+
+                if (logger.isTraceEnabled()) {
+                    logger.trace("Prepare new Prometheus metric collector for: [{}], [{}], [{}]", clusterName, nodeId, nodeName);
+                }
+                PrometheusMetricsCollector collector = new PrometheusMetricsCollector(clusterName, nodeName, nodeId);
```
Comment (vvanholl) [inline src/main/java/org/elasticsearch/rest/action/prometheus/RestPrometheusMetricsAction.java]: Have you checked this doesn't create too overhead ?

Comment (vvanholl) [inline src/main/java/org/elasticsearch/rest/action/prometheus/RestPrometheusMetricsAction.java]: Have you checked this doesn't create to many overhead ?

**Comment by vvanholl** [comment]
For your information the master branch now use Gradle and I have my own Jenkins server which make the builds and the artifacts.
For the moment it is a private part but I want to make it public to allow testing on external PRs.

**Comment by lukas-vlcek** [comment]
> Do you plan porting changes to master 5.x branch ?
> ... the master branch now use Gradle

Yes I do. And I noticed the Gradle and I haven't been using it much in the past but that is life :-) I will try to find a way how to port those pieces of maven magic around integration tests to Gradle.

> I have my own Jenkins server which make the builds and the artifacts. For the moment it is a private part but I want to make it public to allow testing on external PRs.

Well, I guess/hope it should be possible just add another CI service. Or does GitHub support only a single CI service integration? In other words, if GitHub supports more CI services then I do not see any issues if you keep your Jenkins and just add another one for unit/integration tests only.

I also noticed you are using something for reviews (you do not like GitHub reviews?). 

> How and when the async part is triggered ?

Check the `TransportNodePrometheusMetricsAction` class. There are two nested async actions (kind of callback interfaces). And then you can see that `TransportNodePrometheusMetricsAction` is registered to REST action in `PrometheusExporterPlugin` like this:
````
actionModule.registerAction(NodePrometheusMetricsAction.INSTANCE, TransportNodePrometheusMetricsAction.class);
````
And a few supporting classes for serialization of requests and responses are needed.

File: src/main/java/org/elasticsearch/rest/action/prometheus/RestPrometheusMetricsAction.java (hunk 1 above)
Comment (lukas-vlcek) [reply src/main/java/org/elasticsearch/rest/action/prometheus/RestPrometheusMetricsAction.java]: I was thinking about this as well. I was looking into simple java client and it seemed to me that there was just some Hash map created and populated. But I did not run any perf tests. I think we should be able to make some optimizations here as well but I was not sure if this was really worth it given this should ideally processed once per 10 sec or so.

I also wanted to reach out to Prometheus forum about this and ask how the Java client should be used in multithread/parallel environment. There is not much about this topic in their docs.

We can investigate this more if you want.

Comment (lukas-vlcek) [reply src/main/java/org/elasticsearch/rest/action/prometheus/RestPrometheusMetricsAction.java]: Duplicate?

**Comment by davidkarlsen** [comment]
"I have my own Jenkins server which make the builds and the artifacts.
For the moment it is a private part but I want to make it public to allow testing on external PRs." - how about using travis: https://travis-ci.org/ - it's quite versatile and easy, free for opensource, and integrates well with github.

**Comment by vvanholl** [comment]
- You can use both Jenkins and Travis or even Circle CI at the same time. In fact GitHub can handle multiple hooks.

- For reviews I personally rely only on Github, but I noticed some externals PRs are linked to other reviewing systems.

**Comment by lukas-vlcek** [comment]
Thanks @vvanholl

**Comment by jcantrill** [comment]
@lukas-vlcek Does this change exist for 5.5.2.x?  What is required to make it available if it does not?

**Comment by lukas-vlcek** [comment]
It does not. One of the most critical issues have been fixed in this plugin itself and that is this plugin no longer holds and reuses a single ES client for all the requests. This was causing issues with SearchGuard because the ES client is connected with initial request context (meaning it could have been using wrong user for later requests).

But there were other features that this fix implemented and IMO would be very useful to bring them into ES 5.x version. The most useful one is the ability to better control access to the REST API in SearchGuard (ie `cluster:monitor/prometheus/metrics`). But this requires some non-trivial modifications and it is not possible to simply cherry-pick it from 2.x branch for the following reasons:

- ES 5.x is likely to introduce API breaking changes. This shouldn't be hard to figure out. However, to make sure that it works correctly it is critical to have valid unit tests, see below... 
- I implemented integration tests using `ESIntegTestCase`, but it is quite unclear if that would be good approach for ES 5.x and forward, see https://github.com/vvanholl/elasticsearch-prometheus-exporter/issues/66
- ES 5.x switched from Maven to Groovy which I think will have impact on how integration tests needs are executed. And I did not have a time to investigate this in detail.

The other problem I see here is that bringing all this into the plugin will change the code from relatively high-level ES API use to much lower level API (making it more sensitive to internal API changes in ES). And I am not sure we can promise the plugin developers our quick help when they will want release plugin from the most recent ES version while are behind. May be they will not need our help... I do not know, but you know how tricky it can be to implement unit test for ES, right? (Notice, for this plugin it is critical to be able to start several ES nodes in unit tests)

Is there any particular feature you are looking for Jeff for the ES 5.x?

**Pull request merged**